text/0000-app-image-extensions.md

```diff
@@ -21,48 +22,139 @@ Allowing buildpacks to install OS packages dynamically during build would drasti
 
-Extension:
+The buildpack can then exclude directories from the layer by writing to the `launch.toml` using a new `[[excludes]]` table, for example:
+
+```toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
 ```
-/cnb/image/build/extend (build-metadata-toml-file) (cache-dir) | cwd: /
-/cnb/image/run/extend (run-metadata-toml-file) (cache-dir) | cwd: /
+
+* `paths` - a list of path globs to exclude from the run image
+* `cache` - if set to `true`, the paths will be stored in the cache
+
+The `[process]` and `[[slices]]` tables can be used as normal.
+
+The following constraints will be enforced:
+* If a root buildpack is included in a `[[build.buildpacks]]` list in an app's `project.toml`, the build will fail.
+* If a user attempts to create a buildpackage including both root buildpacks and non-root buildpacks, the process will fail.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Building an App with Additional Packages
+
+Given a root buildpack with ID `example/git`, the following configuration can be used:
+
+`project.toml`:
+```toml
+[[stack.build.buildpacks]]
+id = "example/git"
+version = "1.0"
+
+[[stack.run.buildpacks]]
+id = "example/git"
+version = "1.0"
 ```
-Status:
+
+When the following command is run, the app source code will be loaded into the image for access by the root buildpack.
+
+`pack build sclevine/myapp`
+
+This creates a version of the current builder with additional layers and an ephemeral run image with additional layers, then does a normal `pack build`.
+
+* The run image layers are persistented as part of the app image, and can be reused on subsequent builds.
+* The build image layers will be persisted as part of the cache, and can be reused on subsequent builds.
+* Any paths that are excluded using `launch.toml` and set to be cached will be stored in the cache, and can be reused on subsequent builds.
+
+## Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral.
+
+When the base image has not changed
+1. If the buildpacks are configure as idempotent (the default) load the previous layers onto the base images.
+1. Run the buildpacks, creating an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+When there is an update to either the build or run base images
+1. Pull the new base image(s)
+1. Run the root buildpacks against the new image(s) without loading previous layers, and create an emphemeral image.
+1. Rebase the app on to the new emphemeral image.
+
+## Example: Apt Buildpack
+
+A buildpack that reads an `apt.toml` file to install an arbitrary list of packages would have a `buildpack.toml` like this:
```

Comment:
Is it necessary to support running "root buildpacks" with access to application source like this example suggests? 

Requiring access to source feels confusing for run image buildpacks which presumably would not have access to the application source. Additionally, it precludes the option of pre-building builders with the root buildpacks for distribution.

Reply:
we should discuss this in WG. I think it's required for Detect, but idk about Build